### PR TITLE
Update Transformer model and add samples

### PR DESCRIPTION
**Pull request by tbennun**
- [x] Adds a GPT-style causal language model and cleans up the samples
    - [x] Adds parameters such as attention dropout, pre-layernorm, encoder/decoder only stacks, etc.
- [x] Adds `--job-name` to the standard set of scheduler arguments.
- [x] Implements the AdamW weight decay scheme in Adam
- [x] GELU as an operator
- [x] More progress bar information: objective function, print newline every N minibatches

### CODE REVIEW

**Review by ndryden — COMMENTED** [review]
Submitting some comments now, still going over the rest.

File: applications/nlp/transformer/dataset_utils.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        if file.endswith('.py'):
+            result.append(os.path.basename(file)[:-3])
+    return result
+
+
+def load_dataset(name: str):
+    """
+    Loads a dataset by importing the requested module.
+    """
+    sys.path.append(dataset_dir)
```
[the remark this answers is not in the document]
Comment (tbennun) [reply applications/nlp/transformer/dataset_utils.py:28]: We always load one dataset

File: applications/nlp/transformer/datasets/thepile.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                                          split=('train', 'validation'),
+                                          cache_dir=os.path.join(
+                                              data_dir, 'cache'))
+
+# Use the WMT 2016 tokenizer (?)
+tokenizer = Tokenizer.from_file(
+    os.path.join(utils.paths.wmt_dir(), 'tokenizer-wmt16.json'))
+pad_index = tokenizer.token_to_id('<pad>')
+bos_index = tokenizer.token_to_id('<s>')
+eos_index = tokenizer.token_to_id('</s>')
```
[the remark this answers is not in the document]
Comment (tbennun) [reply applications/nlp/transformer/datasets/thepile.py]: Fixed

File: src/optimizers/adam.cpp [hunk 3]
```diff
@@ -206,7 +217,8 @@ void adam<TensorDataType>::step_compute_cpu(AbsDistMatrixType& values,
       auto& m2 = moment2_buffer[i];
       m1 = m_beta1 * m1 + (one - m_beta1) * g;
       m2 = m_beta2 * m2 + (one - m_beta2) * g * g;
-      x -= correction * m1 / (El::Sqrt(m2) + m_eps);
+      x -=
+        correction * (m1 / (El::Sqrt(m2) + m_eps) + m_adamw_weight_decay * x);
```
Comment (ndryden) [inline src/optimizers/adam.cpp:221]: I think this may apply the AdamW weight decay wrong. This scales the weight decay term by `correction`, which is `learning_rate * sqrt(1-beta2) / (1 - beta1)`, and so scales the weight decay by the correction factor for the moments.

I believe it should instead scale the weight decay term only by the learning rate:
```
x -= correction * m1 / (El::Sqrt(m2) + m_eps) + learning_rate * m_adamw_weight_decay * x
```

Please double-check this.

File: src/optimizers/adam.cpp [hunk 4]
```diff
@@ -229,7 +241,8 @@ void adam<TensorDataType>::step_compute_cpu(AbsDistMatrixType& values,
         auto& m2 = moment2_buffer[row + col * moment2_ldim];
         m1 = m_beta1 * m1 + (one - m_beta1) * g;
         m2 = m_beta2 * m2 + (one - m_beta2) * g * g;
-        x -= correction * m1 / (El::Sqrt(m2) + m_eps);
+        x -=
+          correction * (m1 / (El::Sqrt(m2) + m_eps) + m_adamw_weight_decay * x);
```
Comment (ndryden) [inline src/optimizers/adam.cpp:245]: Ditto.

File: src/optimizers/adam.cu [hunk 5]
```diff
@@ -62,7 +63,7 @@ adam_noncontiguous_kernel(size_t height,
     auto& x = values[row + col * values_ldim];
     m1 = beta1 * m1 + (TensorDataType(1) - beta1) * g;
     m2 = beta2 * m2 + (TensorDataType(1) - beta2) * g * g;
-    x -= correction * m1 / (gpu_lib::sqrt(m2) + eps);
+    x -= correction * (m1 / (gpu_lib::sqrt(m2) + eps) + weight_decay * x);
```
Comment (ndryden) [inline src/optimizers/adam.cu:66]: Ditto.

File: src/optimizers/adam.cu [hunk 6]
```diff
@@ -89,7 +91,7 @@ adam_contiguous_kernel(size_t size,
     auto& x = values[gid];
     m1 = beta1 * m1 + (TensorDataType(1) - beta1) * g;
     m2 = beta2 * m2 + (TensorDataType(1) - beta2) * g * g;
-    x -= correction * m1 / (gpu_lib::sqrt(m2) + eps);
+    x -= correction * (m1 / (gpu_lib::sqrt(m2) + eps) + weight_decay * x);
```
Comment (ndryden) [inline src/optimizers/adam.cu:94]: Ditto.

**Review by ndryden — CHANGES_REQUESTED** [review]
Transformer looks good but I think we should double-check AdamW.

File: applications/nlp/transformer/main.py
[the remark this answers is not in the document]
Comment (tbennun) [reply applications/nlp/transformer/main.py:1]: it's similar to a deprecation notice, since others may have used that file before; but I can remove it. What do you think?

File: src/optimizers/adam.cpp (hunk 3 above)
Comment (tbennun) [reply src/optimizers/adam.cpp]: Fixed.

File: src/optimizers/adam.cpp (hunk 4 above)
Comment (tbennun) [reply src/optimizers/adam.cpp]: Fixed

File: src/optimizers/adam.cu (hunk 5 above)
Comment (tbennun) [reply src/optimizers/adam.cu]: Fixed

**Review by ndryden — APPROVED** [review]
Looks good to me.

File: applications/nlp/transformer/parallelism.py
[the remark this answers is not in the document]
Comment (tbennun) [reply applications/nlp/transformer/parallelism.py:1]: Implemented.

**Review by ndryden — APPROVED** [review]
